### ISSUE TITLE
deployment produced an unexpected new value: .traffic_filter: was null, but now...

### ISSUE THREAD

**Issue by tommed** [elastic/terraform-provider-ec#676]
## Readiness Checklist

- [x] I am running the latest version
- [x] I checked the documentation and found no answer
- [x] I checked to make sure that this issue has not already been filed
- [x] I am reporting the issue to the correct repository (for multi-repository projects)

## Expected Behavior

I have attempted to deploy an deployment which doesn't have any traffic filters using Terraform.

## Current Behavior

```
When applying changes to ec_deployment.example_minimal, provider
"provider[\"registry.terraform.io/elastic/ec\"]" produced an unexpected new
value: .traffic_filter: was null, but now
cty.SetVal([]cty.Value{cty.StringVal("b888ca7f774f495f9ec756fc334ddea9")}).

This is a bug in the provider, which should be reported in the provider's own
issue tracker.
Error: Process completed with exit code 1.
```

## Terraform definition

```tf

data "ec_stack" "latest" {
  version_regex = "latest"
  region        = "gcp-europe-west2"
}

# create deployment instance
resource "ec_deployment" "example_minimal" {
  name = "Test 123"

  region                 = "gcp-europe-west2"
  version                = data.ec_stack.latest.version
  deployment_template_id = "gcp-storage-optimized-dense"

  traffic_filter = [
  ]

  elasticsearch = {
    hot = {
      autoscaling = {}
    }
  }

  kibana = {}
}
```

## Steps to Reproduce

1. Use the configuration above
2. Run the configuration
3. Results in the error included above

## Context

It's not possible to deploy the deployment. If i change the traffic filters to an empty array, it then (incorrectly) attempts to delete and recreate the entire deployment, but still results in the same issue.

## Possible Solution

None. This doesn't appear to be useable.

## Your Environment

* Version used: 0.7.0
* Running against Elastic Cloud SaaS or Elastic Cloud Enterprise and version: latest
* Environment name and version (e.g. Go 1.9): `ubuntu-latest`
* Server type and version: GitHub Actions
* Operating System and version: Ubuntu
* Link to your project: *private repo*

**Comment by tobio** [comment]
Is the `b888ca7f774f495f9ec756fc334ddea9` traffic filter configured to be automatically applied to all deployments? It would explain why this filter is being applied to a deployment freshly created via TF.

I'd expect this to be fixed already in master by https://github.com/elastic/terraform-provider-ec/pull/632. @dimuon is there an indication on when the keystore changes will be ready to make a new release here?

**Comment by tommed** [comment]
Are you saying if I change the version in here, it'll work because this is already fixed?

```tf
terraform {
  required_version = ">= 0.12.29"

  required_providers {
    ec = {
      source  = "elastic/ec"
      version = ">= 0.7.0"
    }
  }
}
```

**Comment by tobio** [comment]
> Are you saying if I change the version in here, it'll work because this is already fixed?

No, I'm saying that it's been fixed, but not yet released. You could theoretically build the provider yourself from master, but I'm intentionally leaving that as an exercise. 

If `b888ca7f774f495f9ec756fc334ddea9` is configured to be applied to all deployments then IIUC you should just be able to set `traffic_filter = ["b888ca7f774f495f9ec756fc334ddea9"]` on the ec_deployment resource and have everything Just Work:tm:

**Comment by tommed** [comment]
Ok, so same as the ticket: https://github.com/elastic/terraform-provider-ec/issues/433#issuecomment-1600590623
Is there any ETA, as for us we feel this provider isn't really viable to use - particularly as it tries to delete the database instance if any configuration changes occur - deleting a live production database would be bad! 🤣

**Comment by tommed** [comment]
Ok understood. Sounds like to release cycle is pretty long for terraform providers.. Please don't take this the wrong way, but respectfully I would ask that you guys try and be sympathetic to people who are using your products in production environments and are answerable to their own SLAs. We are reliant of your efforts because we've been told Terraform is 'the way to go' and have put our trust and reliance in you. 

It's hard to explain to customers they have to wait weeks or months for something to work. For example, #433 was opened back in February and makes the EC provider totally unusable. Appreciate the hard work, Thanks.

**Comment by efabens** [comment]
> @tobio , I created PRs for both ECE API and the provider but still need some time for IT tests. Also I will need to update Go client SDK after ECE API PR's merge. I'll try to complete this in the following 2 weeks taking into account my sidetrack tasks.

@dimuon can you link to those PRs, so that I may track when they are merged, and update my provider version promptly.

I am running into a very similar issue, returning the same error. In this case the `ec_deployment` is claiming to own filters rules where it shouldn't. All filters associated with the deployment were linked using `ec_deployment_traffic_filter_association` resources. But at plan/apply time the deployment tries to remove them, and fails. 

```
resource "ec_deployment" "cluster" {
  region                 = var.region
  name                   = "elastic-gcp-deployment"
  version                = data.ec_stack.latest.version
  deployment_template_id = "gcp-storage-optimized-v5"
  alias                  = "<removed>"

  elasticsearch = {
    hot = {
      autoscaling = {}
    }
  }
  kibana = {}
}

resource "ec_deployment_traffic_filter" "home_filter" {
  name   = "home_filters"
  region = var.region
  type   = "ip"
  rule {
    source = "<removed>"
    description = "<removed>"
  }
  rule {
    source      = "<removed>"
    description = "<removed>"
  }
}

# Note: This hardcoded deployment id is a workaround for a bug in the provider. It must be deployed using a `--target` flag.
# Since the addition of traffic filters, the elastic deployment cannot be reapplied using terraform..️
# a fix appears to be in final testing 🤞 https://github.com/elastic/terraform-provider-ec/issues/676
resource "ec_deployment_traffic_filter_association" "squirrel_nests" {
  deployment_id     = "<removed>"
  traffic_filter_id = ec_deployment_traffic_filter.home_filter.id
}
```
```
│ Error: Provider produced inconsistent result after apply
│
│ When applying changes to ec_deployment.cluster, provider "provider[\"registry.terraform.io/elastic/ec\"]" produced an unexpected new value: .traffic_filter: was null, but now
│ cty.SetVal([]cty.Value{cty.StringVal("<removed>"), cty.StringVal("<removed>"), cty.StringVal("<removed>"),
│ cty.StringVal("<removed>"), cty.StringVal("<removed>"), cty.StringVal("<removed>"), cty.StringVal("<removed>")}).
│
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
```

Thank you!

**Comment by tobio** [comment]
@efabens [this](https://github.com/elastic/terraform-provider-ec/pull/674) is the provider PR we're waiting on for the next release. 

I'm closing this issue out as it's been fixed in master, and will be part of the next release.

**Comment by tommed** [comment]
Do we have an ETA on it being released please?

**Comment by tobio** [comment]
This got released with 0.8.0 yesterday.
